algo/src/main/java/org/neo4j/graphalgo/impl/metaPathComputation/ComputeAllMetaPaths.java

```diff
@@ -145,58 +127,36 @@ private void computeMetaPathsFromAllNodeLabels() {
         }
     }
 
-    private void computeMetaPathFromNodeLabel(ArrayList<Integer> pCurrentMetaPath, HashMap<Integer, Integer> pCurrentInstances, int pMetaPathLength) {
-        Stack<ArrayList<Integer>> param1 = new Stack();
-        Stack<HashMap<Integer, Integer>> param2 = new Stack();
-        Stack<Integer> param3 = new Stack();
-        param1.push(pCurrentMetaPath);
-        param2.push(pCurrentInstances);
-        param3.push(pMetaPathLength);
-
-        ArrayList<Integer> currentMetaPath;
-        HashMap<Integer, Integer> currentInstances;
-        int metaPathLength;
-
-        while(!param1.empty() && !param2.empty() && !param3.empty())
-        {
-            currentMetaPath = param1.pop();
-            currentInstances = param2.pop();
-            metaPathLength = param3.pop();
-
-            if (metaPathLength <= 0) {
-                continue;
-            }
+    private void computeMetaPathFromNodeLabel(ArrayList<Integer> currentMetaPath, HashMap<Integer, Integer> currentInstances, int metaPathLength) {
+        if (metaPathLength <= 0) {
+            return;
+        }
 
-            ArrayList<HashMap<Integer, Integer>> nextInstances = allocateNextInstances();
-            fillNextInstances(currentInstances, nextInstances);
+        ArrayList<HashMap<Integer, Integer>> nextInstances = calculateNextInstances(currentInstances);
 
-            for (int edgeLabel : arrayGraphInterface.getAllEdgeLabels()) {
-                for (int nodeLabel : arrayGraphInterface.getAllLabels()) {
-                    int key = labelDictionary.get(new AbstractMap.SimpleEntry<>(edgeLabel, nodeLabel));
-                    HashMap<Integer, Integer> nextInstancesForLabel = nextInstances.get(key);
-                    if (!nextInstancesForLabel.isEmpty()) {
-                        nextInstances.set(key, null);
-
-                        ArrayList<Integer> newMetaPath = copyMetaPath(currentMetaPath);
-                        newMetaPath.add(edgeLabel);
-                        newMetaPath.add(nodeLabel);
-
-                        long instanceCountSum = 0;
-                        for (int count : nextInstancesForLabel.values()) {//refactor to stream?
-                            instanceCountSum += count;
-                        }
-
-                        addAndLogMetaPath(newMetaPath, instanceCountSum);
-
-                        param1.push(newMetaPath);
-                        param2.push(nextInstancesForLabel);
-                        param3.push(metaPathLength - 1);
+        for (int edgeLabel : arrayGraphInterface.getAllEdgeLabels()) {
+            for (int nodeLabel : arrayGraphInterface.getAllLabels()) {
+                int instancesArrayIndex = labelDictionary.get(new AbstractMap.SimpleEntry<>(edgeLabel, nodeLabel));
```

Comment:
Pair? And so on. there are some comments concerning this, but I did not mark all appearances of SimpleEntry though.